iroha_p2p/src/network.rs

```diff
@@ -93,114 +125,207 @@ where
         // from peer
         self.broker.subscribe::<PeerMessage<T>, _>(ctx);
         // from other iroha subsystems
         self.broker.subscribe::<Post<T>, _>(ctx);
+        // to be able to stop all of this
+        self.broker.subscribe::<StopSelf, _>(ctx);
+
+        let (sender, receiver) = oneshot::channel();
+        self.finish_sender = Some(sender);
         // register for peers from listener
         #[allow(clippy::expect_used)]
         let listener = self
             .listener
             .take()
             .expect("Unreachable, as it is supposed to have listener on the start");
-        ctx.notify_with_context(Self::listener_stream(listener, self.public_key.clone()));
+        ctx.notify_with_context(Self::listener_stream(
+            listener,
+            self.public_key.clone(),
+            receiver,
+        ));
     }
 }
 
 #[async_trait::async_trait]
 impl<T, K, E> Handler<Connect> for NetworkBase<T, K, E>
 where
-    T: Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
+    T: Debug + Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
     type Result = ();
 
-    async fn handle(&mut self, Connect { id }: Connect) {
-        match Peer::new_to(id.clone(), self.broker.clone()) {
-            Ok(peer) => {
-                let peer = peer.start().await;
-                self.peers.insert(id, peer);
+    async fn handle(&mut self, Connect { mut id }: Connect) {
+        info!(
+            "[{}] Creating new peer actor for {:?}",
+            &self.listen_addr, &id
+        );
+        id.public_key = self.public_key.clone();
+        let peer = match Peer::new_to(id, self.broker.clone()) {
+            Ok(peer) => peer,
+            Err(e) => {
+                warn!(%e, "Unable to create peer");
+                return;
             }
-            Err(e) => warn!(%e, "Unable to create peer"),
-        }
+        };
+
+        let connection_id = peer.connection_id;
+        let peer = peer.start().await;
+        debug!("Inserting {} into new_peers", connection_id);
+        self.new_peers.insert(connection_id, peer.clone());
+        peer.do_send(Start).await;
     }
 }
 
 #[async_trait::async_trait]
 impl<T, K, E> Handler<Post<T>> for NetworkBase<T, K, E>
 where
-    T: Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
+    T: Debug + Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
     type Result = ();
 
     async fn handle(&mut self, msg: Post<T>) {
-        let addr: &Addr<Peer<T, K, E>> = &self.peers[&msg.id];
-        addr.do_send(msg).await;
+        match self.peers.get(&msg.id.public_key) {
+            Some(Connection(addr, _)) => addr.do_send(msg).await,
+            None if msg.id.public_key == self.public_key => debug!("Not sending message to myself"),
```

Comment:
We should not have debug code in our main codebase. If you want to warn in case of erroneous behavior use `iroha_logger::warn`